Sprint 12

SelickSD opened the pull request.

SelickSD committed:
Сделал загрузку фотографий из сети

SelickSD committed:
Сделал общую функцию для URLRequest

SelickSD committed:
Сделал загрузку фото из сети
Спринт 12_Загрузка картинок из сети

SelickSD committed:
Сделал функционал с лайками

SelickSD committed:
Сделал переход к синглвью через тап по картинке

SelickSD committed:
Сделал логаут

SelickSD committed:
Поменял  dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'" на ISO8601DateFormatter() согласно требований задания

SelickSD committed:
Исправил алерт и убрал лишние комиты

andrew-kuzn-yndx commented on YaImageFeed/Services/ImagesListService.swift:
```diff
@@ -0,0 +1,128 @@
+    private (set) var photos: [Photo] = []
+    private var lastLoadedPage: Int?
+    static let DidChangeNotification = Notification.Name(rawValue: "ImagesListServiceDidChange")
+    private static let BATCH_SIZE = 10
```
Не забывай, что мы используем нейминг с маленькой буквы и по правилу camelCase :)

andrew-kuzn-yndx commented on YaImageFeed/Services/ImagesListService.swift:
```diff
@@ -0,0 +1,128 @@
+    func fetchPhotosNextPage() {
+
+
+        let nextPage = 1 + (Int(photos.count) / ImagesListService.BATCH_SIZE)
```
Было бы здорово разделять блоки кода пустой строкой. Так код выглядит более читаемо и аккуратно. SwiftLint (сервис для чистого кода) например, тоже рекомендует разделять блоки кода одной пустой строкой.

andrew-kuzn-yndx commented on YaImageFeed/ViewControllers/ImagesListViewController.swift:
```diff
@@ -6,43 +6,123 @@
         super.viewDidLoad()
 
+        imagesListServiceObserver = NotificationCenter.default
+            .addObserver(
```
При добавлении Observer, нужно предусматривать и его удаление
Это нужно реализовать в других этапах жизненного цикла контроллера обязательно

andrew-kuzn-yndx commented on YaImageFeed/ViewControllers/ImagesListViewController.swift:
```diff
@@ -6,43 +6,123 @@
+        photos = imagesListService.photos
+
+        if oldCount != newCount {
+            tableView.performBatchUpdates {
```
Тут необходимо прописывать [weak self], чтобы избежать цикла сильных ссылок. 
Можно взять за правило - что всегда, при обращении к self. из замыкания, нужно писать [weak self], чтобы код был безопасным)

andrew-kuzn-yndx commented on YaImageFeed/ViewControllers/ImagesListViewController.swift:
```diff
@@ -72,13 +152,39 @@ extension ImagesListViewController: UITableViewDataSource {
+        print(#function)
+        guard let indexPath = tableView.indexPath(for: cell) else { return }
+        let photo = photos[indexPath.row]
+        imagesListService.changeLike(photoId: photo.id, isLike: photo.isLiked, index: indexPath.row) {result in
```
Снова забыл [weak self]

andrew-kuzn-yndx commented on YaImageFeed/ViewControllers/SingleImageViewController.swift:
```diff
@@ -39,8 +45,51 @@ final class SingleImageViewController: UIViewController {
+        let alert: [UIAlertAction] = [
+            UIAlertAction(title: "Не надо", style: .default, handler: { _ in
+            }),
+            UIAlertAction(title: "Повторить", style: .default, handler: { _ in
```
[weak self]

andrew-kuzn-yndx commented on YaImageFeed/ViewControllers/SingleImageViewController.swift:
```diff
@@ -39,8 +45,51 @@ final class SingleImageViewController: UIViewController {
+                    rescaleAndCenterImageInScrollView(image: imageResult.image)
+
+                case .failure(_):
+                    UIBlockingProgressHUD.dismiss()
```
Дублирование UIBlockingProgressHUD. Его можно вывести за пределы switch, так как его вызывать в любом случае)

SelickSD committed:
Сделал корректировки по замечаниям ревьювера

Добавил алерт на нажатие кнопки выхода из аккаунта
Добавил кастомный активити индикатор
Добавил удаление Observer

SelickSD committed:
Убрал пробелы

SelickSD committed:
Добавил в сториборд настройку плейсхолдера и корнер радиус

SelickSD committed:
Перенес настройки ячейки в cellForRowAt
